### ISSUE TITLE
GSSoC: Implement User Model and Authentication Routes (Signup/Signin) with Zod Validation

### ISSUE THREAD

**Issue by himanshuraimau** [oxiton-foundation/click-metrics#89]
**Description**:
We need to set up user authentication for our backend project. This involves creating a user model and implementing the necessary routes for user signup and signin. Additionally, we will use Zod for input validation to ensure data integrity and security.

**Tasks**:
1. **Create User Model**:
    - Define a User schema using Mongoose.
    - Include fields for username, email, password, and any other relevant information.
    - Add necessary validations (e.g., email format, password length).

2. **Setup Authentication Routes**:
    - **Signup Route**:
        - Endpoint: `POST /api/auth/signup`
        - Accepts user details (username, email, password).
        - Validates input using Zod.
        - Creates a new user in the database.
        - Returns a success message or error response.
    - **Signin Route**:
        - Endpoint: `POST /api/auth/signin`
        - Accepts email and password.
        - Validates input using Zod.
        - Authenticates user credentials.
        - Returns a JWT token on successful authentication or an error response.

3. **Hash Passwords**:
    - Use bcrypt to hash passwords before saving to the database.

4. **Environment Variables**:
    - Add necessary environment variables (e.g., JWT secret, bcrypt salt rounds).

**Proposed Solution**:
- Use Mongoose for database modeling.
- Use bcrypt for password hashing.
- Use JSON Web Tokens (JWT) for authentication.
- Use Zod for input validation to ensure proper data formats and security.

**Request for Assignment**:
I would like to work on this feature as part of GSSoC. Could you please assign this issue to me?

Thank you!

**Comment by rahulpoonia29** [comment]
@himanshuraimau You can work on the issue

**Comment by ManpreetSingh18** [comment]
@Swarnendu0123 can you please assign me this issue as no one is working on this? I have earlier also implemented this authentication. You can check out my work at [tavpe]( https://tavpe.vercel.app/ ).

**Comment by himanshuraimau** [comment]
actually he said to use firebase so use that !

**Comment by ManpreetSingh18** [comment]
In his message, mongoose is written and in your code, I have seen mongo DB used .

**Comment by ManpreetSingh18** [comment]
I agree with you but as per #189 opened there it is mentioned explicitly to use it in this way. So I meet all the requirements as mentioned in the issue itself.
